catalog/app/components/Preview/loaders/Notebook.js

```diff
@@ -16,3 +23,76 @@ export const Loader = function NotebookLoader({ handle, children }) {
   )
   return children(utils.useErrorHandling(processed, { handle, retry: data.fetch }))
 }
+
+const IFRAME_SANDBOX_ATTRIBUTES = 'allow-scripts allow-same-origin'
+const IFRAME_LOAD_TIMEOUT = 30000
+
+function waitForIframe(src) {
+  let resolved = false
+
+  return new Promise((resolve, reject) => {
+    const handleError = (error) => {
+      document.body.removeChild(link)
+      reject(error)
+    }
+
+    const handleSuccess = () => {
+      resolved = true
+      document.body.removeChild(link)
+      resolve(src)
+    }
+
+    const timerId = setTimeout(() => {
+      if (resolved) return
+      handleError(new Error('Page is taking too long to load'))
+    }, IFRAME_LOAD_TIMEOUT)
+
+    const link = document.createElement('iframe')
+    link.addEventListener('load', () => {
+      clearTimeout(timerId)
+      handleSuccess()
+    })
+    link.src = src
+    link.style.display = 'none'
+    link.sandbox = IFRAME_SANDBOX_ATTRIBUTES
+
+    document.body.appendChild(link)
+
+    const iframeDocument = link.contentWindow || link.contentDocument
+    if (iframeDocument) {
+      iframeDocument.addEventListener('error', handleError)
+    }
+  })
+}
+
+async function loadVoila({ src }) {
+  // Preload iframe, then insert cached iframe
+  await waitForIframe(src)
+  return PreviewData.IFrame({ src, sandbox: IFRAME_SANDBOX_ATTRIBUTES })
+}
+
+const useVoilaUrl = (handle) => {
+  const sign = AWS.Signer.useS3Signer()
+  const endpoint = Config.use().registryUrl
+  return React.useMemo(
+    () => `${endpoint}/voila/voila/render/${mkSearch({ url: sign(handle) })}`,
+    [endpoint, handle, sign],
```

Comment:
with this approach we should make sure that `handle` object identity doesnt change, otherwise the url will be regenerated every time new handle object is created, even if its props are unchanged. we can either pass individual props to `useMemo` dependencies (like `[handle.bucket, handle.key, ...`) or use `useMemoEq` since it leverages deep equality checks